### PR TITLE
Fix embed suppression on non-self users when cache is enabled

### PR DESCRIPTION
**Pull request by iulianalexa**
This pull request proposes a fix for issue #2644.

### CODE REVIEW

**Review by GnomedDev — CHANGES_REQUESTED** [review]
This is a bad solution, a proposed solution would be to move the check into `EditMessage::execute`, pass along the `author_id` in the Context tuple, and check there while having access to the message flags to check if suppress embeds has been set.

**Comment by iulianalexa** [comment]
> This is a bad solution, a proposed solution would be to move the check into `EditMessage::execute`, pass along the `author_id` in the Context tuple, and check there while having access to the message flags to check if suppress embeds has been set.

I agree that a more organised solution involves doing the check in execute, however my solution checks the other fields of the struct too in addition to the suppress embeds flag, the thought process being that a request with suppress embeds and a content change would still be invalid.

I would like to know if I should keep using that approach (create a reference struct, check the serialised equality) or if I should just check the flag and let the Discord API handle it.

**Comment by GnomedDev** [comment]
Why not just compare them directly? Deriving PartialEq might be needed, but it would be a ton more efficent than passing through serde.

**Comment by iulianalexa** [comment]
If I compare them directly, then is it not fine to do it in `edit` like in the initial proposal, but without the serialisation? Otherwise, I would have to do the cache check inside `execute`, which to me does not make sense considering the original author check was done inside of `edit`. Furthermore, there is one more function calling `execute` that does not use a context (`ChannelId::edit_message`), so I would have to wrap it inside an `Option`, further complicating things and making them look not so nice.

To me it's fine either way, but I thought I'd let you know before I push it.

**Comment by iulianalexa** [comment]
I have implemented the proposed solution, please let me know if anything should be changed.

File: src/builder/edit_message.rs [hunk 1]
```diff
@@ -254,6 +254,19 @@ impl Builder for EditMessage {
     ) -> Result<Self::Built> {
         self.check_length()?;
 
+        if let Some(user_id) = ctx.2 {
+            #[cfg(feature = "cache")]
```
Comment (alexandruradovici) [inline src/builder/edit_message.rs]:
```suggestion
        #[cfg(feature = "cache")]
        if let Some(user_id) = ctx.2 {
```

I think check should not be done at all if the cache feature is not enabled.

File: src/builder/edit_message.rs [hunk 2]
```diff
@@ -254,6 +254,19 @@ impl Builder for EditMessage {
...
+        if let Some(user_id) = ctx.2 {
+            {
+                if let Some(cache) = cache_http.cache() {
+                    let reference_builder = EditMessage::new().suppress_embeds(true);
+
+                    if user_id != cache.current_user().id && self != reference_builder {
+                        return Err(Error::Model(ModelError::InvalidUser));
+                    }
+                }
+            }
```
Comment (alexandruradovici) [inline src/builder/edit_message.rs]:
```suggestion
                if let Some(cache) = cache_http.cache() {
                    let reference_builder = EditMessage::new().suppress_embeds(true);

                    if user_id != cache.current_user().id && self != reference_builder {
                        return Err(Error::Model(ModelError::InvalidUser));
                    }
                }
```

You can delete the inner scope.

Comment (iulianalexa) [reply src/builder/edit_message.rs]: Thank you for the suggestion

**Comment by alexandruradovici** [comment]
Please fix the formatting issue.

**Review by arqunis — APPROVED** [review]

**Pull request merged**